### PR TITLE
Add reenigne's improved CGA composite patch

### PR DESCRIPTION
**Pull request by kcgen**
This combines reenigne's CGA composite patch, kindly maintained by VileRancour, with NewRisingSun's already merged PCjr composite patch. 

PCjr composite now uses these improvements too:  colors, hue, contrast, brightness, and sharpness. 

This is currently a draft for functional testing and early preview. Feedback on the merge or testing results is very welcome. 

To get this merged will we still need to apply the usual Staging adjustments with a couple more commits; but those should be mostly topical.

After this PR, we can add conf setting for composite , because manually "turning the dials" every game launch is tedious, and we want preservationists to lock-in these preferences that best suit each game (ie: grass is green, sky is blue, etc..).

### CODE REVIEW

File: src/hardware/vga_other.cpp [hunk 1]
```diff
@@ -641,13 +675,14 @@ static void PCJr_FindMode()
 			/* bit3 of mode control 2 signals 2 colour graphics mode */
 			if (cga_comp == 1 ||
 			    (cga_comp == 0 && !(vga.tandy.mode_control & 0x4))) {
-				VGA_SetMode(M_CGA16);
+				VGA_SetMode(M_CGA2_COMPOSITE);
```
Comment (kcgen) [inline src/hardware/vga_other.cpp:678]: Likely needs update_color_16() here too.

**Comment by kcgen** [comment]
This is amazing work by @reengine. The artists and coders at Sierra maximized the results from CGA's 160x200x16 mode on circa-1985 hardware!

Early-model CGA Composite:



Late-model CGA Composite:

**Comment by kcgen** [comment]
And without the scanline shader (old and new):

**Comment by kcgen** [comment]
Let's wrap this up!

@bmunger , @kklobe , @LowLevelMahn -- appreciate any feedback and help you can offer: code review / testing / and so on.

The first commit is the big one - this simply merges reengine's patch. The next commits are the usual small staging adjustments.  Suggest reviewing the commits one-at-a-time, as they're broken down into individual themes.  If there are any regressions, the commits are fine-grained enough that hopefully we can isolate the problem without having to rollback this PR entirely.

That said, it's passing all of the video-mode tests I can throw at it (not only composite), but EGA/VGA/SVGA/etc.  UBSAN and ASAN builds also run clean.

To test it, use this: [composite-test-pack.zip](https://github.com/dosbox-staging/dosbox-staging/files/6890179/composite-test-pack.zip)

 - King's Quest 2: CGA composite is automatically used by the game.
 - BC Quest for Tire: press F12 to switch from TLL to CGA composite mode
 - Seven Cities of Gold: press F12 to switch to PCjr composite mode
 - Ultima II: press F12 to switch to PCjr composite mode

Once in composite mode:
 - You can select a CRT knob to adjust by pressing F10 (which cycles through the knobs).
 - Turn the selected knob with F11 or (Alt+F11).
 
the PR also fixes 24 issues flagged by PVS studio:

File: src/hardware/vga_draw.cpp [hunk 2]
```diff
@@ -73,37 +73,123 @@ static Bit8u * VGA_Draw_2BPPHiRes_Line(Bitu vidstart, Bitu line) {
 	return TempLine;
 }
 
-static Bitu temp[643]={0};
+static Bit8u byte_clamp(int v) {
+	v >>= 13;
+	return v < 0 ? 0 : (v > 255 ? 255 : v);
+}
```
Comment (kklobe) [inline src/hardware/vga_draw.cpp:89]: future note: C++17 has `std::clamp()`

Comment (kcgen) [reply src/hardware/vga_draw.cpp:89]: This is a very unique clamp that divides the inbound value by 8192 and then clamps the result.

Comment (kklobe) [reply src/hardware/vga_draw.cpp:89]: yep, was thinking `return v < 0 ? 0 : (v > 255 ? 255 : v);` could become eg. `return std::clamp(v,0,255);`

Comment (kcgen) [reply src/hardware/vga_draw.cpp:89]: (d'oh.. indeed!)

**Review by kklobe — COMMENTED** [review]
Verified output with included composite pack, and especially like the cleanup converting macro functions into constexpr ones.

**Comment by kcgen** [comment]
Thanks for the review @kklobe!
Also very nice to hear it brought back memories of your original Roland composite display via PCjr output; testament to the painstaking work by **reengine**!
I think we're ready to merge this.

**Pull request merged**